### PR TITLE
Save and resume training from an intermediate checkpoint

### PR DESCRIPTION
**Pull request by amajee11us**
Having 8 L40S GPUs (paper requirements) can become a challenge to most users (including me) and can take a few days to complete end-to-end finetuning (as tested on 2xH100 GPUs with the same setting as the paper). This requires a checkpointing and resuming pipelining in place. Following are the changes -

1. Saving model checkpoint - Save the model weights (only trainable params) with the optimizer and scheduler states.
2. Loading model checkpoint - Similar to 1 load only the trainable weights and optimizer states. Also the current changelist skips the number of steps and resumes the model loading from there. 

The changes are in response to #7

### CODE REVIEW

**Comment by tyleryzhu** [comment]
Hi amajee, 

Apologies for the very delayed response on this. I just started a new job so that was hectic, but finally got some time to look more in depth. Appreciate the changes, this looks good esp. the FSDP loading!
